### DISCUSSION TITLE
Gas Estimation using ethers.js

### DISCUSSION THREAD

**Discussion by cpspk** [community/community#142202] [Programming Help] [answered]
### Body

Hi web3 developer.
How are you today?
I am developing my own project that can interact with smart contracts.
But I have an blocker while estimating the gas using ethers.js
Here is my current codebase.
```
    const signer = await wallet.getSigner();
    const contract = new ethers.Contract(contractAddress, abi, signer);
    const gasPrice = await getGasPrice();
    const gasEstimate = await contract.estimateGas.createSession(
      newSession.sessionName,
      newSession.sessionCreator,
      newSession.rewardToken,
      parseInt(newSession.priceShare),
      parseInt(newSession.duration),
      parseInt(newSession.maxParticipants),
      parseInt(newSession.minParticipants),
      allowedParticipants,
      newSession.referrerAddress,
      isPrivate,
      { value: ethers.parseEther("0.01") }
    );
    const gasFee = gasEstimate * gasPrice! + protocolFee
```
I just followed the documentation of ethers.js but not working.
Please help me.

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by sven3270350** [comment]
Which version of ethers.js are you using now?

**Reply by cpspk** [reply]
now I am using version 6.

**Reply by sven3270350** [reply, marked as answer]
Then you can call the estimateGas function like this.
```
    const gasEstimate = await contract.createSession.estimateGas(
      newSession.sessionName,
      newSession.sessionCreator,
      newSession.rewardToken,
      parseInt(newSession.priceShare),
      parseInt(newSession.duration),
      parseInt(newSession.maxParticipants),
      parseInt(newSession.minParticipants),
      allowedParticipants,
      newSession.referrerAddress,
      isPrivate,
      { value: ethers.parseEther("0.01") }
    );
```

**Reply by cpspk** [reply]
One more question please.
Can I call the read function on the contract with out the wallet signer?

**Reply by sven3270350** [reply]
Yes, You can call the read functions without any signer.
The signer is needed only for write contract.

**Reply by cpspk** [reply]
Really thanks for your help.

**Reply by sven3270350** [reply]
You are welcome!

**Comment by Aninooo** [comment]
Check the Function Signature // Make sure that the function createSession in your smart contract actually accepts the parameters you’re passing and that they are of the correct types.

Use BigNumber // Ethers.js recommends using BigNumber for handling large numbers. When calculating the gas fee, ensure you use BigNumber for both the gas estimate and the gas price.

Handle Promises Properly // Ensure that all asynchronous calls (like getting the gas price) are properly awaited.

**Comment by cpspk** [comment]
https://github.com/orgs/community/discussions/142202#discussioncomment-11010328
